armory/adapt/pytorch.py

```diff
@@ -0,0 +1,1126 @@
+#         #    pass
+#
+#     def early_stop(self):
+#         pass
+#
+#     def early_stop_inner(self):
+#         pass
+#
+#     def init(self, x, i=0):
+#         """
+#         Initialize starting point
+#
+#         i is the current step and is ignored in the base version
+#         """
+#         if self.random:
+#             pass
+#         self.x = x
+#
+#     def get_gradient(self):
+#         return self.estimator.get_grad()
+#
+#     # def project(self, x):
+#     #    return self.project_l2(x, self.x_orig, self.epsilon)
+#
+#     def project(self, x, norm=None):
+#         if norm is None:
+#             norm = self.norm
+#         norm = str(norm)
+#         try:
+#             projection = self.projections[norm]
+#         except KeyError:
+#             raise ValueError(f"norm {norm} is not in {list(self.PROJECTIONS) + [None]}")
+#         return projection(x, self.x_orig)
+#
+#     def clip(self, x):
+#         return x
+#
+#     # TODO: insert pytorch optimizers (e.g., Adam)
+
+
+# class Attack:
+#     def __init__(self, model: torch.nn.Module, domain=None, loss_fn=None):
+#         if not isinstance(model, torch.nn.Module):
+#             raise ValueError(f"{model} is not a torch.nn.Module")
+#         self.model = model
+#         self.loss_fn = loss_fn
+#
+#     def __call__(self, x_orig: torch.Tensor, y_true=None) -> torch.Tensor:
+#
+#         pass
+
+
+# def PGD_linf(model, epsilon=8/255, project=project_linf, clip=(0, 1), pert=linf, random_init=random_linf_init, Optimizer):
+#     pass
+
+
+# ThreatModel = namedtuple("ThreatModel", ["project", "init", "random", "dist", "norm", "clip"])
+# L2ThreatModel = ThreatModel(
+#     project=project_l2,
+#     init=None,
+#     random=None,
+#     dist=l2_dist,
+#     norm=l2,
+#     clip=None,
+# )
+
+
+class ProbLoss(nn.NLLLoss):
+    """
+    Cross Entropy Loss with probability inputs
+    """
+
+    def forward(self, input: Tensor, target: Tensor) -> Tensor:
+        return super().forward(torch.log(input), target)
+
+
+class PGD:
+    def __init__(
+        self,
+        model,
+        epsilon=8 / 255,
+        eps_step=1 / 255,
+        min_clip=0,
+        max_clip=1,
+        random=None,
+        project=None,
+        distance=None,
+        normalize=None,
+        output="prob",
+    ):
+        self.model = model
+        self.proxy = model
+        self.epsilon = epsilon
+        self.eps_step = eps_step
+
+        # TODO: specify these in init
+        self.output_map = {
+            "logit": nn.CrossEntropyLoss(),
+            "logprob": nn.NLLLoss(),
+            "prob": ProbLoss(),
+        }
+        if output not in self.output_map:
+            raise ValueError(f"output {output} not in {list(self.output_map)}")
+        self.output = output
+        self.loss_fn = self.output_map[self.output]
+
+        # logits:
+        self.task_metric = lambda y_pred, y_true: (
+            y_pred.detach().argmax(dim=1) == y_true.detach()
```

Comment:
just pointing out a challenge here. While setting this task_metric is simple enough for classification, this would break for non-classification tasks and is much less straightforward to define in those cases

Reply:
Point taken